### ISSUE TITLE
roachprod: azure gc is broken in specific cases

### ISSUE THREAD

**Issue by arulajmani** [cockroachdb/cockroach#58668]
**Describe the problem**

TLDR: If there is an error when creating the first VM of a cluster, all other resources created for that cluster (storage, network) will not be GC'd and will be left dangling. Worse yet, they won't be GC'd by the periodic GC job that removes expired clusters. They will also be ignored when introspecting using `roachprod list`. The only way to remove the resources would be through the Azure portal's UI.

Unlike GCP and AWS where all resources are tied to VMs, Azure constructs resources (network resources, disk etc.) under Resource Groups and then attaches them to VMs. This means that unlike the other two clouds, even if no VM exists, there may still be resources that need to be cleaned up on failure. 

When trying to destroy a cluster (through the roachprod command, or on an error during the creation process, or periodic GC-ing of expired clusters), roachprod first accumulates all resources using https://github.com/cockroachdb/cockroach/blob/c2f1f7e78664e10e43c77094a7ef6af388e76f0f/pkg/cmd/roachprod/cloud/cluster_cloud.go#L161 
This accumulation is done using `Provider.List()`, which is part of an interface implemented for each cloud. In the case of Azure, this returns a list of all existing VMs. Then, the cluster name (which is also the name of the Resource Group) is derived from the VM name and the following Azure specific deletion code is used to destroy the Resource Group:
https://github.com/cockroachdb/cockroach/blob/c2f1f7e78664e10e43c77094a7ef6af388e76f0f/pkg/cmd/roachprod/vm/azure/azure.go#L241

This is unfortunately not resilient to VM creation failure. If no VM is created under a Resource Group, the accumulation logic described above fails to identify the (now dangling) Resource Group. A possible solution here would be to change the way `Provider.List()` works for Azure -- instead of listing all VMs and then deriving the Resource Group's name in the calling function (from the VM's name), we should instead be simply listing all Resource Groups. 

Lastly, it is worth noting that because this GC bug is caused by not accumulating resources correctly, `roachprod list` wouldn't show these dangling resources either. The only way to find (and delete) them is through the Azure portal.

**To Reproduce**

Create a cluster on Azure using roachprod such that the very first VM creation would fail. One sure shot way would be to create it for a machine type in a region where the creation would exceed quota limits.

**Expected behavior**
Resources (network, disk) should be correctly cleaned up if there is an error when creating the first VM of a roachprod Azure cluster.

Epic CRDB-10428

Jira issue: CRDB-3373

**Comment by arulajmani** [comment]
cc @kenliu @miretskiy

**Comment by arulajmani** [comment]
> A possible solution here would be to change the way Provider.List() works for Azure -- instead of listing all VMs and then deriving the Resource Group's name in the calling function (from the VM's name), we should instead be simply listing all Resource Groups.

Looking at this a bit more, and this may not be feasible, because we still want a list of all VMs that are part of the cluster in other places. Instead, we probably need to remove the assumption that a cluster exists iff a VM under it exists in the code. This only affects Azure. The specific change would be to use the Resource Groups API to enumerate all clusters, and continue using `Provider.List()` the way it is implemented today to nest VMs under Resource Groups.
